storage: refactor leader lease acquisition

andreimatei opened the pull request:
This is done through a regular LeaderLease raft request, naming the new
leader replica.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6999)

<!-- Reviewable:end -->

andreimatei commented:
@tschottdorf does this look like the right direction?

tschottdorf commented:
Can't see the actual changes you made due to the code movement. Just makes this a pain to look at, even if the refactors are probably sensible. I have no way of tracing the changes on a granular level without going above and beyond. Could you separate out your stuff and the refactors pretty please?

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> @tschottdorf does this look like the right direction?
</details>


---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KJMj7NeYsUJ5tyIIu21:1465052102)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: move requestLeaderLease to its own file

To be refactored next.

andreimatei committed:
storage: refactor leader lease acquisition

Create a specialized structure for joining in-progress leadership
requests.  This refactor will make it possible to send leadership
requests crowning a different replica as the leader.

andreimatei commented:
Tobi, please take another look at this. r4-r6 are current.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Can't see the actual changes you made due to the code movement. Just makes this a pain to look at, even if the refactors are probably sensible. I have no way of tracing the changes on a granular level without going above and beyond. Could you separate out your stuff and the refactors pretty please?
</details>


---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KKBNFWUJeXqa56fJ0fp:365413673)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on storage/replica.go:
```diff
@@ -495,12 +496,14 @@ func loadLeaderLease(eng engine.Reader, rangeID roachpb.RangeID) (*roachpb.Lease
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	return r.mu.leaderLease, len(r.mu.llChans) > 0
+	nextLease := r.mu.pendingLeaderLeaseRequest.RequestPending()
```
Why assign?

tschottdorf commented on storage/replica.go:
```diff
@@ -519,32 +522,56 @@ func loadGCThreshold(eng engine.Reader, rangeID roachpb.RangeID) (roachpb.Timest
+// grant the lease is pending. Leases are eagerly renewed when a request with a
+// timestamp close to the beginning of the stasis period is served.
 //
 // TODO(spencer): implement threshold regrants to avoid latency in
```
Remove this TODO (it's already finished).

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+			}
+		}
+		return c.Err
+	case <-r.store.Stopper().ShouldDrain():
```
Add a comment on l.166 that once we use a real context there, we can't do this if we don't manage to abandon the command (see `addWriteCmd`).

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+
+// executeLeaderLeaseCommand proposes a LeaderLeader and waits for its
+// completion.
+func (r *Replica) executeLeaderLeaseCommand(
```
s/execute/propose/

tschottdorf replied:
sorry, probably better `s/execute/send/` - `execute` sounds wrong (but `propose` doesn't quite hit the nail on the head as well).

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+		},
+	}
+	if !replica.store.Stopper().RunAsyncTask(func() {
+		pErr := replica.executeLeaderLeaseCommand(leaseReq, timestamp, startKey)
```
I think you should pass `leaseReq.Lease` instead, then you can construct `leaseReq` in that method (and you wouldn't have to pass the timestamp).

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+// InitOrJoinRequest executes a LeaderLease command asynchronously and returns a
+// promise for the result. If there's already a request in progress, we check
+// that the request is asking for a lease for the same replica as the one this
+// call wants to crown leader. If so, we join in waiting the results of that
```
I would avoid the word `leader`. In fact, I think we should be dropping the word `leader` from `leader lease` altogether (not suggesting you do that now, but anticipating that in the comments would be good).

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+}
 
+// InitOrJoinRequest executes a LeaderLease command asynchronously and returns a
+// promise for the result. If there's already a request in progress, we check
```
it returns a `chan *Error`, which isn't exactly a promise (at least not for the result).

tschottdorf commented on storage/replica.go:
```diff
@@ -519,32 +522,56 @@ func loadGCThreshold(eng engine.Reader, rangeID roachpb.RangeID) (roachpb.Timest
+// It returns true if the current lease covers the timestamp, false if it
+// doesn't. If the lease is not owned, a NotLeaderError is returned.
+// The lease is also extended if owned and almost expiring.
+func (r *Replica) checkLeaseLocked(timestamp roachpb.Timestamp) (bool, *roachpb.Error) {
```
`check` does not suggest taking action when there is no lease.

I don't see the point in factoring this out into its own method. There are enough methods with `lease` in their names to be confusing, and it's only used from one of the other ones from which you extracted it.

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -20,100 +20,164 @@
+			p.llChans = append(p.llChans, llChan)
+			return llChan
+		}
+		llChan <- roachpb.NewErrorf("request for different replica in progress "+
```
This is a change in semantics and not a refactoring. Why is this in this commit? This will return errors all the way to the users if hit. Presumably we don't exercise this code yet, but the point remains.

tschottdorf commented on storage/replica_leader_lease.go:
```diff
@@ -128,6 +135,33 @@ func (p *pendingLeaderLeaseRequest) InitOrJoinRequest(
+	replicaID roachpb.ReplicaID,
+	leaseTransferStasisDuration time.Duration,
+) (bool, *roachpb.Lease) {
+	if nextLease := p.RequestPending(); nextLease != nil {
```
I'm uncomfortable with this. You're relying on in-memory status to preserve correctness, but a quick node restart is possible.

tschottdorf commented:
I suggest opening a PR for the first three commits since they're not related to this PR and can merge (after comments), and leaving this to discuss the remainder once that has merged.

andreimatei commented:
ok, kept only the refactoring here
will send a different PR for the transfer after I get rid of the transfer stasis

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I suggest opening a PR for the first three commits since they're not related to this PR and can merge (after comments), and leaving this to discuss the remainder once that has merged.
</details>


---

Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 505 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r5-505:-KKFmTufC5zpgjkznwKU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica.go#L505)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why assign?
</details>
Done.

---

*[storage/replica.go, line 544 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r5-544:-KKFmvb4WH7oBkClYxSK:1178713008) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica.go#L544)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `check` does not suggest taking action (but it does extend the lease).
> 
> 
> 
> I don't see the point in factoring this out into its own method. There are enough methods with `lease` in their names to be confusing, and it's only used from one of the other ones from which you extracted it.
</details>
how about `checkLeaseAndMaybeExtendLocked`?

I think the extraction helps - the function from which this was extracted was pretty convoluted, with a top level loop and all. Plus this function is called from an inline fun (introduced for defer unlock).
I'd keep it.

---

*[storage/replica.go, line 576 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r5-576:-KKFneNlLf-Fndq-qEVg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica.go#L576)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Remove this TODO (it's already finished). 
</details>
Done.

---

*[storage/replica_leader_lease.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-56:-KKFnhty93irdCBmXAoH:1671819238) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L56)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> it returns a `chan *Error`, which isn't exactly a promise (at least not for the result).
</details>
I think it qualifies as a promise... The error, or lack thereof, is the only result.

---

*[storage/replica_leader_lease.go, line 58 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-58:-KKFo9G1tvUUGdvyWde3:52716221) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L58)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I would avoid the word `leader`. In fact, I think we should be dropping the word `leader` from `leader lease` altogether (not suggesting you do that now, but anticipating that in the comments would be good).
</details>
I've `s/leader lease/lease`. But here where I just say leader, how should I replace it?
Do you want me to replace the file and class names?
Why are we hating on "leader"?

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKJptTC_MZD4Suimm9K:682122386) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This is a change in semantics and not a refactoring. Why is this in this commit? This will return errors all the way to the users if hit. Presumably we don't exercise this code yet, but the point remains. Could you move that one commit over?
</details>
well it's not a change in semantics if the code is not used :)
what do you suggest this commit to do if `nextLeader` doesn't match what's in progress? Or do you want to not pass `nextLeader` yet at all?
Can we just leave it here? Some element of anticipation is fine...

---

*[storage/replica_leader_lease.go, line 100 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-100:-KKJs8HoTPsy6mLUCjf_:1420754924) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L100)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think you should pass `leaseReq.Lease` instead, then you can construct `leaseReq` in that method (and you don't have to pass the timestamp).
</details>
I've inlined the execute fun.

---

*[storage/replica_leader_lease.go, line 152 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-152:-KKK9Dhq9ANaxffv1_vR:677677860) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L152)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> sorry, probably better `s/execute/send/` - `execute` sounds wrong (but `propose` doesn't quite hit the nail on the head as well).
</details>
gone

---

*[storage/replica_leader_lease.go, line 180 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-180:-KKKAKOA75wHaGgu-6H_:1711767949) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L180)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add a comment on l.166 that once we use a real context there, we can't do this if we don't manage to abandon the command (see `addWriteCmd`).
</details>
not sure I follow. Once we use a real context, we won't be able to do what exactly? Return the `NotLeaderError` early? Why can't we return that regardless of whether we manage to abandon the command or not?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KKKN6tR5Y8EWW114N3E:631418679)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 641 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKjNpc68acVu60ftKR:-KKKjNpc68acVu60ftKS:903228120) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L641)):*
> ```Go
> 
> 			// Otherwise, no active lease: Request renewal if a renewal is not already pending.
> 			return r.getOrExtendLeaderLeaseLocked(timestamp), nil
> ```

Why do we (potentially) extend in two different places? It's confusing that checkLeaseAndMaybeExtendLocked will extend the lease if it does cover the given timestamp, but won't try to acquire the lease if it's free.

---

*[storage/replica_leader_lease.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-56:-KKKla-l_xhI9Dz9--Lu:-1915823786) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think it qualifies as a promise... The error, or lack thereof, is the only result.
</details>
The new lease is also a result; I wouldn't refer to this as a "promise" for that reason. 

---

*[storage/replica_leader_lease.go, line 58 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-58:-KKKkc_LFXb4ZVs3Y8VJ:-1258204442) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L58)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I've `s/leader lease/lease`. But here where I just say leader, how should I replace it?
> Do you want me to replace the file and class names?
> Why are we hating on "leader"?
</details>
"Leader lease" is confusingly similar to "raft leader". I've started calling it the "range lease". (one day we'll have quorum leases and there will be more than one lease holder per range)

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKKpFj5p-TfsVPozDQ5:715740065) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well it's not a change in semantics if the code is not used :)
> what do you suggest this commit to do if `nextLeader` doesn't match what's in progress? Or do you want to not pass `nextLeader` yet at all?
> Can we just leave it here? Some element of anticipation is fine...
> 
</details>
Do lease transfers even need to be integrated with `pendingLeaderLeaseRequest`? If it's just for testing then we can do the transfer with a one-off and leave the coalescing mechanisms unaware of it.

---

*[storage/replica_leader_lease.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKs5UlRia1harI9_7N:-KKKs5UmnKzKM8xj38Wn:135520664) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L79)):*
> ```Go
> 			return llChan
> 		}
> 		llChan <- roachpb.NewErrorf("request for different replica in progress "+
> ```

Isn't this a special case of NotLeaderError, instead of a non-retryable unknown error? 

---

*[storage/replica_leader_lease.go, line 164 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKL0ByYq1jbIMxrMf-v:-KKL0ByYq1jbIMxrMf-w:-1611344490) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L164)):*
> ```Go
> // Unless an error is returned, the obtained lease will be valid for a time
> // interval containing the requested timestamp.
> func (r *Replica) getOrExtendLeaderLeaseLocked(timestamp hlc.Timestamp) <-chan *roachpb.Error {
> ```

Why "orExtend"? It looks like this is just a "get".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on storage/replica.go:
```diff
@@ -558,36 +560,59 @@ func loadGCThreshold(eng engine.Reader, rangeID roachpb.RangeID) (hlc.Timestamp,
+) error {
 	err := &roachpb.NotLeaderError{}
 	if l != nil && l.Replica.ReplicaID != 0 {
+		desc := rangeDesc
```
pointless assignment, just rename on to the other.

tschottdorf commented:
Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 544 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r5-544:-KKLCs1MV77G8MXi3zuR:-562337563) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica.go#L544)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about `checkLeaseAndMaybeExtendLocked`?
> 
> I think the extraction helps - the function from which this was extracted was pretty convoluted, with a top level loop and all. Plus this function is called from an inline fun (introduced for defer unlock).
> I'd keep it.
</details>
Yes, I see that point but as @bdarnell too pointed out, putting stuff into a new function != added clarity. If you don't find a good way to separate the concerns, I prefer that you keep it inlined.

---

*[storage/replica.go, line 613 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLDHXd_Vl5JHfApPfN:-KKLDHXd_Vl5JHfApPfO:-585879549) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L613)):*
> ```Go
> // replica currently holds the lease, redirects by returning NotLeaderError. If
> // the lease is expired, a renewal is synchronously requested. This method uses
> // the pendingLeaderLeaseRequest structure to guarantee only one request to
> ```

s/leader// in all of these comments and structs.

---

*[storage/replica.go, line 635 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLDncVZvQlbi8TzRd3:-KKLDncVZvQlbi8TzRd4:-727935746) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L635)):*
> ```Go
> 			}
> 			if covering {
> 				return nil, nil
> ```

Returning the `nil` channel as saying "all is well" deserves an inline comment here as well.

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKLEHyH6ERC_bzU3dro:605870869) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Do lease transfers even need to be integrated with `pendingLeaderLeaseRequest`? If it's just for testing then we can do the transfer with a one-off and leave the coalescing mechanisms unaware of it.
</details>
Again, this code doesn't belong into this PR. This code is problematic if introduced in practice, and introducing it now with the argument that it's unused (but still not wanting to test it) is silly. In this PR, the desired next leader is always the current replica and so you don't need to pass the parameter.

---

*[storage/replica_leader_lease.go, line 180 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-180:-KKLEzjJucJrqUq6r8i0:-2093390204) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L180)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> not sure I follow. Once we use a real context, we won't be able to do what exactly? Return the `NotLeaderError` early? Why can't we return that regardless of whether we manage to abandon the command or not?
</details>
See `addWriteCmd`. Once you return here, any trace in the context will be finished and if the command *does* apply later, it will panic (at least as long as we try to be memory efficient).

---

*[storage/replica_leader_lease.go, line 101 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLErNuLC6HAtu22GNe:-KKLErNv-ACnFjgrl3BK:-1039409159) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L101)):*
> ```Go
> 	}
> 	if !replica.store.Stopper().RunAsyncTask(func() {
> 		// Proposed a LeaderLease command and wait for it to apply.
> ```

s/Proposed/Propose/

---

*[storage/replica_leader_lease.go, line 164 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKL0ByYq1jbIMxrMf-v:-KKLFZmyz7vCXICCVcIp:-547668545) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L164)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why "orExtend"? It looks like this is just a "get".
</details>
The timestamp could during the active lease, in which case this is an extension. But I agree that the name is confusing - the real extension logic is in that other function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 4 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 544 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r5-544:-KKPuF3No0UdfuOxConB:467769323) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica.go#L544)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yes, I see that point but as @bdarnell too pointed out, putting stuff into a new function != added clarity. If you don't find a good way to separate the concerns, I prefer that you keep it inlined.
</details>
ok, removed

---

*[storage/replica.go, line 568 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oL-r8-568:-KKPq3vF9QomroUivJEH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L568)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> pointless assignment, just rename on to the other.
</details>
Done.

---

*[storage/replica.go, line 613 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLDHXd_Vl5JHfApPfN:-KKPzfygPhD0u3nF0LZj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L613)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/leader// in all of these comments and structs.
</details>
Done.

---

*[storage/replica.go, line 635 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLDncVZvQlbi8TzRd3:-KKPzhBkXPO38d6tjJj4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L635)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Returning the `nil` channel as saying "all is well" deserves an inline comment here as well.
</details>
Done.

---

*[storage/replica.go, line 641 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKjNpc68acVu60ftKR:-KKPtprjdbZr3tCVLFYQ:434435167) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica.go#L641)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why do we (potentially) extend in two different places? It's confusing that checkLeaseAndMaybeExtendLocked will extend the lease if it does cover the given timestamp, but won't try to acquire the lease if it's free.
</details>
ok, I got rid of `checkLeaseAndMaybeExtendLocked` and inlined it in the caller.

---

*[storage/replica_leader_lease.go, line 56 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-56:-KKPW7XeKJptWerHVgah:1177676377) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L56)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The new lease is also a result; I wouldn't refer to this as a "promise" for that reason. 
</details>
well, technically, the new lease is a side-effect, not a result of this function. But ok, replaced with "returns a channel to be used for waiting for the results of applying the command". ?

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKPTdB8cLwCNXRbXk5D:-341482145) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Again, this code doesn't belong into this PR. This code is problematic if introduced in practice, and introducing it now with the argument that it's unused (but still not wanting to test it) is silly. In this PR, the desired next leader is always the current replica and so you don't need to pass the parameter.
</details>
@bdarnell I think lease transfers need to be integrated with this `pendingLeaderLeaseRequest`. A transfer is a `LeaderLease` request, just like a lease extension. We shouldn't allow a lease acquisition or extension while a transfer is in progress, and the other way around.

@tschottdorf As discussed, let's just keep these lines here. I've tried to remove it, but the code becomes uglier if you don't pass in a `ReplicaDescriptor`. And if you do pass in a `ReplicaDescriptor`, this code serves as a sanity assertion.

---

*[storage/replica_leader_lease.go, line 180 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-180:-KKQ6lDcEP8YJC2n4lPa:1740160800) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L180)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> See `addWriteCmd`. Once you return here, any trace in the context will be finished and if the command *does* apply later, it will panic (at least as long as we try to be memory efficient).
</details>
added a method comment on this method about it

---

*[storage/replica_leader_lease.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKs5UlRia1harI9_7N:-KKPVqqMfyYGHHmjNVr4:1271985603) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L79)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Isn't this a special case of NotLeaderError, instead of a non-retryable unknown error? 
</details>
well for the purposes of this commit, this should never trigger; it's an assertion.

Once we get transfers, this will be a `NotLeaderError` if there's a transfer in progress and a non-retryable error if there isn't (so non-retryable if you request a transfer while an extension is in progress. `NotLeaderError` would not be appropriate there). 

---

*[storage/replica_leader_lease.go, line 101 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKLErNuLC6HAtu22GNe:-KKPpAlRBhFVgSP_UFQG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L101)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Proposed/Propose/
</details>
Done.

---

*[storage/replica_leader_lease.go, line 164 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKL0ByYq1jbIMxrMf-v:-KKQ6GHbwxeG6Ynqi07w:1367546961) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L164)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The timestamp could during the active lease, in which case this is an extension. But I agree that the name is confusing - the real extension logic is in that other function.
</details>
renamed to `requestLeaseLocked`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 0 of 4 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKQNie2jGlS2GK_x3tz:-958081126) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> @bdarnell I think lease transfers need to be integrated with this `pendingLeaderLeaseRequest`. A transfer is a `LeaderLease` request, just like a lease extension. We shouldn't allow a lease acquisition or extension while a transfer is in progress, and the other way around.
> 
> @tschottdorf As discussed, let's just keep these lines here. I've tried to remove it, but the code becomes uglier if you don't pass in a `ReplicaDescriptor`. And if you do pass in a `ReplicaDescriptor`, this code serves as a sanity assertion.
</details>
Why shouldn't we allow a lease acquisition while a transfer is in progress? One will win; one will fail; there is no guarantee that it will work even without this synchronization because another node could win. And if this is test-only it doesn't really matter. 

---

*[storage/replica_leader_lease.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKs5UlRia1harI9_7N:-KKQOHaghYR3JXlB6WD6:2106104017) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L79)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well for the purposes of this commit, this should never trigger; it's an assertion.
> 
> Once we get transfers, this will be a `NotLeaderError` if there's a transfer in progress and a non-retryable error if there isn't (so non-retryable if you request a transfer while an extension is in progress. `NotLeaderError` would not be appropriate there). 
</details>
That sounds like more reason to leave this code out of this PR if it's going to change before it's actually used.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KKQOdu2rtDC6s8tV_MO:676834233)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 4 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKQSndycS5HhVdFQ3jZ:2069256323) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why shouldn't we allow a lease acquisition while a transfer is in progress? One will win; one will fail; there is no guarantee that it will work even without this synchronization because another node could win. And if this is test-only it doesn't really matter. 
</details>
As soon as a node initiates a transfer, it must stop serving reads. So, this information (that the current lease is tainted) needs to be stored somewhere - and having the transfer integrated in the `pendingLeaseRequest` serves this purpose. This doesn't automatically have to mean that the node can't request a new lease concurrently with the transfer (which would probably fail), but it'd be more code.
Also, as soon as the node has initiated a transfer, it's probably a good thing to speculatively assume it'll succeed and redirect everybody to the next leader, right?

The other way around, supporting a transfer while an extension is in progress, sounds like it could be good. Except I think it'd also complicate the code, since now even if you get a new lease, this new one can also be tainted. And I was planning on relying on a single lease being tainted - the one active when the transfer is initiated - and I was going to identify it somehow in state persisted across node restarts in a node-local range.

So I dunno, having `pendingLeaseRequest` serialize lease accesses sounds convenient to me. If you find this unconvincing or have some good ideas, ping me.

---

*[storage/replica_leader_lease.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKs5UlRia1harI9_7N:-KKQUHXwWPtBPhwICFvB:-628177864) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L79)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> That sounds like more reason to leave this code out of this PR if it's going to change before it's actually used.
</details>
I didn't tell the whole truth - it'll probably be the callers that check whether something is in progress and return a corresponding error. This code will stay as a sanity check assertion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKUFP85xmYzEJBCoIuT:1706302107) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> As soon as a node initiates a transfer, it must stop serving reads. So, this information (that the current lease is tainted) needs to be stored somewhere - and having the transfer integrated in the `pendingLeaseRequest` serves this purpose. This doesn't automatically have to mean that the node can't request a new lease concurrently with the transfer (which would probably fail), but it'd be more code.
> Also, as soon as the node has initiated a transfer, it's probably a good thing to speculatively assume it'll succeed and redirect everybody to the next leader, right?
> 
> The other way around, supporting a transfer while an extension is in progress, sounds like it could be good. Except I think it'd also complicate the code, since now even if you get a new lease, this new one can also be tainted. And I was planning on relying on a single lease being tainted - the one active when the transfer is initiated - and I was going to identify it somehow in state persisted across node restarts in a node-local range.
> 
> So I dunno, having `pendingLeaseRequest` serialize lease accesses sounds convenient to me. If you find this unconvincing or have some good ideas, ping me.
</details>
I thought we decided not to shorten the leases, so the old lease is still good until its natural expiration, and the transferee can't start serving anything yet. If we are shortening leases, then integrating it here makes sense. 

---

*[storage/replica_leader_lease.go, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKKs5UlRia1harI9_7N:-KKUEPuShrfrIZtxJxnM:-81926003) ([raw file](https://github.com/cockroachdb/cockroach/blob/2294ec19f90c202c9e468ed05f2d35c1280da957/storage/replica_leader_lease.go#L79)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I didn't tell the whole truth - it'll probably be the callers that check whether something is in progress and return a corresponding error. This code will stay as a sanity check assertion.
</details>
Like I said, sounds like all the more reason to leave this code out until we can see it its real uses. Remove `nextLeaseHolder` and hard-code it to this replica for now. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r1, 8 of 8 files at r3, 2 of 4 files at r5, 7 of 7 files at r6, 11 of 11 files at r7, 5 of 5 files at r9, 4 of 4 files at r10.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KKUY9g3VqCyCVx5CxBF:1417088721)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[storage/replica_leader_lease.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KJDPd_xWcrrMYE1P9oN-r5-78:-KKUXUZk1qzgm91Q8CKd:-937612424) ([raw file](https://github.com/cockroachdb/cockroach/blob/a40e8b7325e8a1075445bffcc0fe856e9c654d33/storage/replica_leader_lease.go#L78)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I thought we decided not to shorten the leases, so the old lease is still good until its natural expiration, and the transferee can't start serving anything yet. If we are shortening leases, then integrating it here makes sense. 
</details>
we are indeed going to shorten the current lease. More specifically, the new lease that will be proposed for the transfer starts immediately, and the current lease is not to be used any more, also effective immediately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->

tamird commented:
nit: plz wrap commit messages at 72 columns.

---

Reviewed 4 of 4 files at r1, 8 of 8 files at r3, 2 of 2 files at r4, 2 of 4 files at r5, 7 of 7 files at r6, 11 of 11 files at r7, 5 of 5 files at r9, 3 of 4 files at r10.
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 243 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKUdocPFpw0leABApAd:-KKUdocPFpw0leABApAe:1449268233) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L243)):*
> ```Go
> 		// pendingLeaseRequest is used to coalesce LeaderLease requests.
> 		pendingLeaseRequest pendingLeaseRequest
> ```

why the space here?

---

*[storage/replica.go, line 570 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFHd2_8A-5HdUarmh:-KKWFHd2_8A-5HdUarmi:1567063915) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L570)):*
> ```Go
> 	for attempt := 1; ; attempt++ {
> 		timestamp := r.store.Clock().Now()
> 		llChan, pErr := func() (<-chan *roachpb.Error, *roachpb.Error) {
> ```

`llChan` for a channel of errors?

---

*[storage/replica.go, line 590 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWF0pG37xzmPXZ0C5p:-KKWF0pG37xzmPXZ0C5q:-1403454381) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L590)):*
> ```Go
> 					// to go through and simply ignore the returned channel (which
> 					// is buffered).
> 					_ = r.requestLeaseLocked(timestamp)
> ```

nit: why bother assigning

---

*[storage/replica.go, line 703 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFSh2CE-lhPs-kIL6:-KKWFSh2CE-lhPs-kIL7:593557823) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L703)):*
> ```Go
> // errors are returned.
> func (r *Replica) GetReplica() (*roachpb.ReplicaDescriptor, error) {
> 	_, replica := r.Desc().FindReplica(r.store.StoreID())
> ```

why did you unextract this local?

---

*[storage/replica_range_lease.go, line 40 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFZ1N1bt3Bgaawjqb:-KKWFZ1N1bt3Bgaawjqc:1203552058) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica_range_lease.go#L40)):*
> ```Go
> 	// Slice of channels to send on after lease acquisition.
> 	// If empty, then no request is in progress.
> 	llChans []chan *roachpb.Error
> ```

this name is really not very descriptive 

---

*[storage/replica_test.go, line 1037 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWEX1NVmFFiGgXTCJ_:-KKWEX1NVmFFiGgXTCJa:-1938505814) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica_test.go#L1037)):*
> ```Go
> 	}
> 	tc.rng.mu.Lock()
> 	pErr := <-tc.rng.requestLeaseLocked(tc.clock.Now())
> ```

seems like you could add a non-locked variant

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-:-KKUdl4BcJUHmq26Wd2e:-1483660370)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 4 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 243 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKUdocPFpw0leABApAd:-KKjuVZgxpP5l0UfZNTF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why the space here?
</details>
Done.

---

*[storage/replica.go, line 570 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFHd2_8A-5HdUarmh:-KKjteF7OgTsPGVYTLk0:-1743888159) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L570)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `llChan` for a channel of errors?
</details>
yeah... the name was used before for these channels. The important thing about them is the side-effects that have happened by the time you get a value (i.e. a new leader lease), not so much the error on the channel.

---

*[storage/replica.go, line 590 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWF0pG37xzmPXZ0C5p:-KKjsyZ_1kTElYbebcA2:-860939064) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L590)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: why bother assigning
</details>
to make it more clear that there's a result we're ignoring; to tie it to the comment above.

---

*[storage/replica.go, line 703 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFSh2CE-lhPs-kIL6:-KKjtcjl5dS58yYgb-Hs:-1187005728) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica.go#L703)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why did you unextract this local?
</details>
undone

---

*[storage/replica_range_lease.go, line 40 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWFZ1N1bt3Bgaawjqb:-KKju33AkbHGCXIpKhxx:346404762) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica_range_lease.go#L40)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this name is really not very descriptive 
</details>
it's the old name. I don't have a better name.

---

*[storage/replica_test.go, line 1037 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999#-KKWEX1NVmFFiGgXTCJ_:-KKjuKC4n4w6oKQGupPK:-1953249369) ([raw file](https://github.com/cockroachdb/cockroach/blob/4825892b31c25a8df5bb9068122fcffbd0b6a61d/storage/replica_test.go#L1037)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like you could add a non-locked variant
</details>
but nobody needs it...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6999)*
<!-- Sent from Reviewable.io -->
